man/pm2.1
.TH PM2 production process manager
.SH Table of contents
.SS Quick start
.RS
.IP \(bu 2
Installation
\[la]#a1\[ra]
.IP \(bu 2
Usage
\[la]#a2\[ra]
.IP \(bu 2
Examples
\[la]#a3\[ra]
.IP \(bu 2
Different ways to launch a process
\[la]#raw-examples\[ra]
.IP \(bu 2
Options
\[la]#a987\[ra]
.RS
.IP \(bu 2
Schema
\[la]#a988\[ra]
.RE
.IP \(bu 2
How to update PM2?
\[la]#update-pm2\[ra]
.RE
.SS Features
.RS
.IP \(bu 2
Transitional state of apps
\[la]#a4\[ra]
.IP \(bu 2
Process listing
\[la]#a6\[ra]
.IP \(bu 2
Automatic restart process based on memory
\[la]#max-memory-restart\[ra]
.IP \(bu 2
Monitoring CPU/Memory usage
\[la]#a7\[ra]
.IP \(bu 2
Logs management
\[la]#a9\[ra]
.IP \(bu 2
Clustering
\[la]#a5\[ra]
.IP \(bu 2
Watch & Restart
\[la]#watch--restart\[ra]
.IP \(bu 2
Reloading without downtime
\[la]#a690\[ra]
.IP \(bu 2
Make PM2 restart on server reboot
\[la]#a8\[ra]
.IP \(bu 2
JSON app declaration
\[la]#json-app-declaration\[ra]
.RS
.IP \(bu 2
Schema
\[la]#a988\[ra]
.RE
.RE
.SS Windows specifics
.RS
.IP \(bu 2
Windows
\[la]#windows\[ra]
.RE
.SS Deployment \- ecosystem.json
.RS
.IP \(bu 2
Getting started with deployment
\[la]#deployment\[ra]
.IP \(bu 2
Deployment options
\[la]#deployment-help\[ra]
.IP \(bu 2
Considerations
\[la]#considerations\[ra]
.IP \(bu 2
Contributing
\[la]#deployment-contribution\[ra]
.RE
.SS Using PM2 programmatically (via API)
.RS
.IP \(bu 2
Simple example
\[la]#programmatic-example\[ra]
.IP \(bu 2
Programmatic API
\[la]#programmatic-api\[ra]
.RE
.SS Specific
.RS
.IP \(bu 2
Specific features
\[la]#a77\[ra]
.IP \(bu 2
Configuration file
\[la]#a989\[ra]
.IP \(bu 2
Enabling Harmony ES6
\[la]#a66\[ra]
.IP \(bu 2
CoffeeScript
\[la]#a19\[ra]
.IP \(bu 2
Testing PM2 on your prod environment
\[la]#a149\[ra]
.IP \(bu 2
JSON app via pipe
\[la]#a96\[ra]
.RE
.SS Knowledge
.RS
.IP \(bu 2
Stateless apps ?
\[la]#stateless-apps\[ra]
.IP \(bu 2
Transitional state of apps
\[la]#a4\[ra]
.IP \(bu 2
Setup PM2 on server: tutorial
\[la]#a89\[ra]
.IP \(bu 2
Logs and PID files
\[la]#a34\[ra]
.IP \(bu 2
Execute any script: What is fork mode ?
\[la]#a23\[ra]
.RE
.SS More
.RS
.IP \(bu 2
Contributing/Development mode
\[la]#a27\[ra]
.IP \(bu 2
Known bugs and workaround
\[la]#a21\[ra]
.IP \(bu 2
They talk about it
\[la]#a20\[ra]
.IP \(bu 2
License
\[la]#a15\[ra]
.RE
.ti 0
\l'\n(.lu'
.SH Quick start
.PP
<a name="a1"/>
.SH Installation
.PP
The latest PM2 stable version is installable via NPM:
.PP
.RS
.nf
$ npm install pm2@latest \-g
.fi
.RE
.PP
If the above fails use:
.PP
.RS
.nf
$ npm install git://github.com/Unitech/pm2#master \-g
.fi
.RE
.PP
<a name="a2"/>
.SH Usage
.PP
Hello world:
.PP
.RS
.nf
$ pm2 start app.js
.fi
.RE
.PP
<a name="a3"/>
.SH Raw Examples
.PP
.RS
.nf
# Fork mode
$ pm2 start app.js \-\-name my\-api # Name process
# Cluster mode
$ pm2 start app.js \-i 0        # Will start maximum processes with LB depending on available CPUs
$ pm2 start app.js \-i max      # Same as above, but deprecated yet.
# Listing
$ pm2 list               # Display all processes status
$ pm2 jlist              # Print process list in raw JSON
$ pm2 prettylist         # Print process list in beautified JSON
$ pm2 describe 0         # Display all informations about a specific process
$ pm2 monit              # Monitor all processes
# Logs
$ pm2 logs               # Display all processes logs in streaming
$ pm2 ilogs              # Advanced termcaps interface to display logs
$ pm2 flush              # Empty all log file
$ pm2 reloadLogs         # Reload all logs
# Actions
$ pm2 stop all           # Stop all processes
$ pm2 restart all        # Restart all processes
$ pm2 reload all         # Will 0s downtime reload (for NETWORKED apps)
$ pm2 gracefulReload all # Send exit message then reload (for networked apps)
$ pm2 stop 0             # Stop specific process id
$ pm2 restart 0          # Restart specific process id
$ pm2 delete 0           # Will remove process from pm2 list
$ pm2 delete all         # Will remove all processes from pm2 list
# Misc
$ pm2 reset <process>    # Reset meta data (restarted time...)
$ pm2 updatePM2          # Update in memory pm2
$ pm2 ping               # Ensure pm2 daemon has been launched
$ pm2 sendSignal SIGUSR2 my\-app # Send system signal to script
$ pm2 start app.js \-\-no\-daemon
.fi
.RE
.SH Different ways to launch a process
.PP
.RS
.nf
$ pm2 start app.js           # Start app.js
$ pm2 start app.js \-\- \-a 23  # Pass arguments '\-a 23' argument to app.js script
$ pm2 start app.js \-\-name serverone # Start a process an name it as server one
                                    # you can now stop the process by doing
                                    # pm2 stop serverone
$ pm2 start app.js \-\-node\-args="\-\-debug=7001" # \-\-node\-args to pass options to node V8
$ pm2 start app.js \-i 0             # Start maximum processes depending on available CPUs (cluster mode)
$ pm2 start app.js \-\-log\-date\-format "YYYY\-MM\-DD HH:mm Z"    # Log will be prefixed with custom time format
$ pm2 start app.json                # Start processes with options declared in app.json
                                    # Go to chapter Multi process JSON declaration for more
$ pm2 start app.js \-e err.log \-o out.log  # Start and specify error and out log
.fi
.RE
.PP
For scripts in other languages:
.PP
.RS
.nf
$ pm2 start echo.pl \-\-interpreter=perl
$ pm2 start echo.coffee
$ pm2 start echo.php
$ pm2 start echo.py
$ pm2 start echo.sh
$ pm2 start echo.rb
.fi
.RE
.PP
The interpreter is set by default with this equivalence:
.PP
.RS
.nf
{
  ".sh": "bash",
  ".py": "python",
  ".rb": "ruby",
  ".coffee" : "coffee",
  ".php": "php",
  ".pl" : "perl",
  ".js" : "node"
}
.fi
.RE
.PP
<a name="a987"/>
.SH Options
.PP
.RS
.nf
Options:
   \-h, \-\-help                           output usage information
   \-V, \-\-version                        output the version number
   \-v \-\-version                         get version
   \-s \-\-silent                          hide all messages
   \-m \-\-mini\-list                       display a compacted list without formatting
   \-f \-\-force                           force actions
   \-n \-\-name <name>                     set a <name> for script
   \-i \-\-instances <number>              launch [number] instances (for networked app)(load balanced)
   \-l \-\-log [path]                      specify entire log file (error and out are both included)
   \-o \-\-output <path>                   specify out log file
   \-e \-\-error <path>                    specify error log file
   \-p \-\-pid <pid>                       specify pid file
   \-\-max\-memory\-restart <memory>        specify max memory amount used to autorestart (in megaoctets)
   \-\-env <environment_name>             specify environment to get specific env variables (for JSON declaration)
   \-x \-\-execute\-command                 execute a program using fork system
   \-u \-\-user <username>                 define user when generating startup script
   \-c \-\-cron <cron_pattern>             restart a running process based on a cron pattern
   \-w \-\-write                           write configuration in local folder
   \-\-interpreter <interpreter>          the interpreter pm2 should use for executing app (bash, python...)
   \-\-log\-date\-format <momentjs format>  add custom prefix timestamp to logs
   \-\-no\-daemon                          run pm2 daemon in the foreground if it doesn't exist already
   \-\-merge\-logs                         merge logs from different instances but keep error and out separated
   \-\-watch                              watch application folder for changes
   \-\-ignore\-watch <folders|files>       folder/files to be ignored watching, chould be a specific name or regex \- e.g. \-\-ignore\-watch="test node_modules "some scripts""
   \-\-node\-args <node_args>              space delimited arguments to pass to node in cluster mode \- e.g. \-\-node\-args="\-\-debug=7001 \-\-trace\-deprecation"
   \-\-no\-color                           skip colors
.fi
.RE
.PP
<a name="update\-pm2"/>
.SH How to update PM2
.PP
Install the latest pm2 version :
.PP
.RS
.nf
$ npm install pm2@latest \-g
.fi
.RE
.PP
Then update the in\-memory PM2 :
.PP
.RS
.nf
$ pm2 updatePM2
.fi
.RE
.SH Features
.PP
<a name="a4"/>
.SH Transitional state of apps (important)
.PP
PM2 is a process manager. PM2 can start, stop, restart and \fIdelete\fP processes.
.PP
Start a process:
.PP
.RS
.nf
$ pm2 start app.js \-\-name "my\-api"
$ pm2 start web.js \-\-name "web\-interface"
.fi
.RE
.PP
Now let's say I need to stop the web\-interface:
.PP
.RS
.nf
$ pm2 stop web\-interface
.fi
.RE
.PP
As you can see \fBthe process hasn't disappeared\fP\&. It's still there but in \fB\fCstopped\fR status.
.PP
To restart it just do:
.PP
.RS
.nf
$ pm2 restart web\-interface
.fi
.RE
.PP
Now I want to \fBdelete\fP the app from the PM2 process list.
To do so:
.PP
.RS
.nf
$ pm2 delete web\-interface
.fi
.RE
.PP
<a name="a6"/>
.SH Process listing
.PP
To list all running processes:
.PP
.RS
.nf
$ pm2 list
# Or
$ pm2 [list|ls|l|status]
.fi
.RE
.PP
To get more details about a specific process:
.PP
.RS
.nf
$ pm2 describe 0
.fi
.RE
.PP
<a name="max\-memory\-restart"/>
.SH Automatic restart process based on memory
.PP
Value passed is in megaoctets. Internally it uses the V8 flag \fB\fC\-\-max\-old\-space\-size=MEM\fR to make a process exit when memory exceed a certain amount of RAM used.
.PP
CLI:
\fB\fCbash
$ pm2 start big\-array.js \-\-max\-memory\-restart 20M
\fR
.PP
JSON:
\fB\fCjson
{
  "name" : "max_mem",
  "script" : "big\-array.js",
  "max_memory_restart" : "20M"
}
\fR
.PP
Units can be 
.BR K (ilobyte), 
.BR M (egabyte), 
.BR G (igabyte).
.PP
.RS
Actually the way it works when you type : 
\fB\fCpm2 start app.js \-\-max\-memory\-restart 50M\fR
commander module will transform it to \fB\fCmaxMemoryRestart: "50M"\fR, then after being processed by PM2 logic it will become an env variable as follows \fB\fCmax_memory_restart : 52428800 // in bytes this time\fR\&.
But since programmatic interface doesn't use commander you have to give it raw\-mode : \fB\fCmaxMemoryRestart\fR\&.
.RE
.PP
From this issue comment
\[la]https://github.com/Unitech/PM2/issues/1159#issuecomment-91615025\[ra]\&.
.PP
<a name="a7"/>
.SH Monitoring CPU/Memory usage
.PP
Monitor all processes launched:
.PP
.RS
.nf
$ pm2 monit
.fi
.RE
.PP
<a name="a9"/>
.SH Logs management
.SS Enable Timestamp Prefix of \fB\fCpm2.log\fR
.PP
.RS
.nf
export PM2_LOG_DATE_FORMAT="YYYY\-MM\-DD HH:mm Z"
.fi
.RE
.PP
If this env\-variable has been changed, you need to dump your processes and kill daemon, restart it again to take effect, e.g.:
.PP
.RS
.nf
$ pm2 dump
$ pm2 [resurrect|save]
.fi
.RE
.SS Displaying logs in realtime
.PP
Displaying logs of specified process or all processes in realtime:
.PP
.RS
.nf
$ pm2 logs
$ pm2 logs big\-api
$ pm2 flush # Clear all the logs
.fi
.RE
.SS Advanced log interface
.PP
Navigate between processes logs in realtime with an ergonomic interface:
.PP
.RS
.nf
$ pm2 ilogs
.fi
.RE
.SS Reloading all logs (SIGUSR2/Logrotate)
.PP
To reload all logs, you can send \fB\fCSIGUSR2\fR to the PM2 process.
.PP
You can also reload all logs via the command line with:
.PP
.RS
.nf
$ pm2 reloadLogs
.fi
.RE
.SS Options
.PP
.RS
.nf
\-\-merge\-logs : merge logs from different instances but keep error and out separated
\-\-log\-date\-format <format>: prefix logs with formated timestamp (http://momentjs.com/docs/#/parsing/string\-format/)
.fi
.RE
.SS Merge \fB\fCout\fR and \fB\fCerr\fR logs
.RS
.IP \(bu 2
If you only want to merge \fB\fCout\fR and \fB\fCerr\fR logs into one output file, try with the following examples:
.RE
.PP
.RS
.nf
  $ pm2 start name.js \-o name.log \-e name.log
.fi
.RE
.PP
.RS
.nf
  {
    "apps" : [{
      "name"        : "name",
      "script"      : "name.js",
      "err_file"    : "name.log",
      "out_file"    : "name.log"
    }]
  }
.fi
.RE
.RS
.IP \(bu 2
How about merge \fB\fCout\fR and \fB\fCerr\fR logs into one, and also keep the separated logs? e.g.:
.RE
.PP
.RS
.nf
  $ pm2 start \-l
.fi
.RE
.PP
.RS
.nf
  $ pm2 start \-l name.log
.fi
.RE
.PP
.RS
.nf
  {
    "apps" : [{
      "name"        : "name",
      "script"      : "name.js",
      "log_file"    : true
    }]
  }
.fi
.RE
.PP
.RS
.nf
  {
    "apps" : [{
      "name"        : "name",
      "script"      : "name.js",
      "log_file"    : "name.log"
    }]
  }
.fi
.RE
.IP
\fBNotes:\fP When you providing a Boolean (\fB\fCtrue\fR) value for \fB\fC\-l, \-\-log\fR option or \fB\fClog_file\fR property, it means a merged log file will be automatic generated, i.e. \fB\fC~/.pm2/logs/[name]\-[id].log\fR, otherwise it's specific.
.PP
<a name="a5"/>
.SH Clustering (cluster_mode)
.PP
The \fIcluster_mode\fP will automatically wrap your Node.js app into the cluster module and will enable you to reload your app without downtime and to scale your processes across all CPUs available.
.PP
To enable the \fIcluster_mode\fP, just pass the \-i <instances> option:
.PP
.RS
.nf
$ pm2 start app.js \-i 1
.fi
.RE
.PP
To launch \fB\fCmax\fR instances (\fB\fCmax\fR depending on the number of CPUs available) and set the load balancer to balance queries among process:
.PP
.RS
.nf
$ pm2 start app.js \-\-name "API" \-i 0
.fi
.RE
.PP
DEPRECATED (STILL COMPATIBLE):
.PP
.RS
.nf
$ pm2 start app.js \-\-name "API" \-i max
.fi
.RE
.PP
If your app is well\-designed (\fBstateless\fP) you'll be able to \fBprocess many more queries\fP\&.
.PP
Important concepts to make a Node.js app stateless:
.RS
.IP \(bu 2
Sessions must not be stored in memory but shared via a database (Redis, Mongo, whatever)
.IP \(bu 2
WebSocket/Socket.io should communicate via a database
\[la]http://socket.io/docs/using-multiple-nodes/#passing-events-between-nodes\[ra]
.RE
.PP
<a name="a690"/>
.SH Reloading without downtime
.PP
As opposed to \fB\fCrestart\fR, which kills and restarts the process, \fB\fCreload\fR achieves a 0\-second\-downtime reload.
.PP
\fBWarning\fP This feature only works for apps in \fIcluster_mode\fP, that uses HTTP/HTTPS/Socket connections.
.PP
To reload an app:
.PP
.RS
.nf
$ pm2 reload api
.fi
.RE
.PP
If the reload system hasn't managed to reload your app, a timeout will simply kill the process and will restart it.
.SS Graceful reload
.PP
Sometimes you can experience a \fBvery long reload, or a reload that doesn't work\fP (fallback to restart).
.PP
It means that your app \fBstill has open connections on exit\fP\&.
.PP
To work around this problem you have to use the graceful reload.
Graceful reload is a mechanism that will send a \fBshutdown\fP message to your process before reloading it.
You can control the time that the app has to shutdown via the \fB\fCPM2_GRACEFUL_TIMEOUT\fR environment variable.
.PP
Example:
.PP
.RS
.nf
process.on('message', function(msg) {
  if (msg == 'shutdown') {
    // Your process is going to be reloaded
    // You have to close all database/socket.io/* connections
    console.log('Closing all connections...');
    // You will have 4000ms to close all connections before
    // the reload mechanism will try to do its job
    setTimeout(function() {
      console.log('Finished closing connections');
      // This timeout means that all connections have been closed
      // Now we can exit to let the reload mechanism do its job
      process.exit(0);
    }, 1500);
  }
});
.fi
.RE
.PP
Then use the command:
.PP
.RS
.nf
$ pm2 gracefulReload [all|name]
.fi
.RE
.PP
When PM2 starts a new process to replace an old one, it will wait for the new process to begin listening to a connection or a timeout before sending the shutdown message to the old one. You can define the timeout value with the \fB\fCPM2_GRACEFUL_LISTEN_TIMEOUT\fR environamente variable. If a script does not need to listen to a connection, it can manually tell PM2 that the process has started up by calling \fB\fCprocess.send('online')\fR\&.
.PP
<a name="a8"/>
.SH Startup script
.PP
PM2 has the amazing ability to \fBgenerate startup scripts and configure them\fP\&.
PM2 is also smart enough to \fBsave all your process list\fP and to \fBbring back all your processes on restart\fP\&.
.PP
.RS
.nf
$ pm2 startup
# auto\-detect platform
$ pm2 startup [platform]
# render startup\-script for a specific platform, the [platform] could be one of:
#   ubuntu|centos|redhat|gentoo|systemd|darwin|amazon
.fi
.RE
.PP
Once you have started the apps and want to keep them on server reboot do:
.PP
.RS
.nf
$ pm2 save
.fi
.RE
.PP
\fBWarning\fP It's tricky to make this feature work generically, so once PM2 has setup your startup script, reboot your server to make sure that PM2 has launched your apps!
.SS More information
.PP
Three types of startup scripts are available:
.RS
.IP \(bu 2
SystemV init script (with the option \fB\fCubuntu\fR or \fB\fCcentos\fR)
.IP \(bu 2
OpenRC init script (with the option \fB\fCgentoo\fR)
.IP \(bu 2
SystemD init script (with the \fB\fCsystemd\fR option)
.RE
.PP
The startup options are using:
.RS
.IP \(bu 2
\fBubuntu\fP will use \fB\fCupdaterc.d\fR and the script \fB\fClib/scripts/pm2\-init.sh\fR
.IP \(bu 2
\fBcentos\fP/\fBredhat\fP will use \fB\fCchkconfig\fR and the script \fB\fClib/scripts/pm2\-init\-centos.sh\fR
.IP \(bu 2
\fBgentoo\fP will use \fB\fCrc\-update\fR and the script \fB\fClib/scripts/pm2\fR
.IP \(bu 2
\fBsystemd\fP will use \fB\fCsystemctl\fR and the script \fB\fClib/scripts/pm2.service\fR
.IP \(bu 2
\fBdarwin\fP will use \fB\fClaunchd\fR to load a specific \fB\fCplist\fR to resurrect processes after reboot.
.RE
.SS User permissions
.PP
Let's say you want the startup script to be executed under another user.
.PP
Just use the \fB\fC\-u <username>\fR option !
.PP
.RS
.nf
$ pm2 startup ubuntu \-u www
.fi
.RE
.SS Related commands
.PP
Dump all processes status and environment managed by PM2:
\fB\fCbash
$ pm2 dump
\fR
It populates the file \fB\fC~/.pm2/dump.pm2\fR by default.
.PP
To bring back the latest dump:
\fB\fCbash
$ pm2 [resurrect|save]
\fR
.SH Watch & Restart
.PP
PM2 can automatically restart your app when a file changes in the current directory or its subdirectories:
.PP
.RS
.nf
$ pm2 start app.js \-\-watch
.fi
.RE
.PP
If \fB\fC\-\-watch\fR is enabled, stopping it won't stop watching:
\- \fB\fCpm2 stop 0\fR will not stop watching
\- \fB\fCpm2 stop \-\-watch 0\fR will stop watching
.PP
Restart toggle the \fB\fCwatch\fR parameter when triggered.
.PP
To watch specific paths, please use a JSON app declaration, \fB\fCwatch\fR can take a string or an array of paths. Default is \fB\fCtrue\fR:
.PP
.RS
.nf
{
  "watch": ["server", "client"],
  "ignore_watch" : ["node_modules", "client/img"],
  "watch_options": {
    "followSymlinks": false
  }
}
.fi
.RE
.PP
As specified in the Schema
\[la]#a988\[ra]:
\- \fB\fCwatch\fR can be a boolean, an array of paths or a string representing a path. Default to \fB\fCfalse\fR
\- \fB\fCignore_watch\fR can be an array of paths or a string, it'll be interpreted by chokidar
\[la]https://github.com/paulmillr/chokidar#path-filtering\[ra] as a glob or a regular expression.
\- \fB\fCwatch_options\fR is an object that will replace options given to chokidar. Please refer to chokidar documentation
\[la]https://github.com/paulmillr/chokidar#api\[ra] for the definition.
.PP
PM2 is giving chokidar these Default options:
.PP
.RS
.nf
var watch_options = {
  persistent    : true,
  ignoreInitial : true
}
.fi
.RE
.SH JSON app declaration
.PP
You can define parameters for your apps in \fB\fCprocesses.json\fR:
.PP
.RS
.nf
{
  "apps" : [{
    "name"        : "echo",
    "script"      : "examples/args.js",
    "args"        : ["\-\-toto=heya coco", "\-d", "1"],
    "log_date_format"  : "YYYY\-MM\-DD HH:mm Z",
    "ignore_watch" : ["[\\\\/\\\\\\\\]\\\\./", "node_modules"],
    "watch"       : true,
    "node_args"   : "\-\-harmony",
    "cwd"         : "/this/is/a/path/to/start/script",
    "env": {
        "NODE_ENV": "production",
        "AWESOME_SERVICE_API_TOKEN": "xxx"
    }
  },{
    "name"       : "api",
    "script"     : "./examples/child.js",
    "instances"  : 4,
    "log_date_format"  : "YYYY\-MM\-DD",
    "log_file"   : "./examples/child.log",
    "error_file" : "./examples/child\-err.log",
    "out_file"   : "./examples/child\-out.log",
    "pid_file"   : "./examples/child.pid",
    "exec_mode"  : "cluster_mode",
    "port"       : 9005
  },{
    "name"       : "auto\-kill",
    "script"     : "./examples/killfast.js",
    "min_uptime" : "100s",
    "exec_mode"  : "fork_mode"
  }]
}
.fi
.RE
.PP
Then run:
\fB\fCbash
$ pm2 start processes.json
$ pm2 stop processes.json
$ pm2 delete processes.json
$ pm2 restart processes.json
\fR
.PP
\fBA few notes about JSON app declarations:\fP
.RS
.IP \(bu 2
All command line options passed when using the JSON app declaration will be dropped i.e.
.RE
.PP
.RS
.nf
$ cat node\-app\-1.json
{
  "name" : "node\-app\-1",
  "script" : "app.js",
  "cwd" : "/srv/node\-app\-1/current"
}
.fi
.RE
.RS
.IP \(bu 2
JSON app declarations are additive.  Continuing from above:
\fB\fCbash
$ pm2 start node\-app\-2.json
$ ps aux | grep node\-app
root  14735  5.8  1.1  752476  83932 ? Sl 00:08 0:00 pm2: node\-app\-1
root  24271  0.0  0.3  696428  24208 ? Sl 17:36 0:00 pm2: node\-app\-2
\fR
Note that if you execute \fB\fCpm2 start node\-app\-2\fR again, it will spawn an additional instance node\-app\-2.
.IP \(bu 2
\fBcwd:\fP your JSON declaration does not need to reside with your script.  If you wish to maintain the 
.BR JSON (s) 
in a location other than your script (say, \fB\fC/etc/pm2/conf.d/node\-app.json\fR) you will need to use the cwd feature.  (Note, this is especially helpful for capistrano style directory structures that use symlinks.)  Files can be either relative to the cwd directory, or absolute (example below.)
.IP \(bu 2
The following are valid options for JSON app declarations:
\fB\fCjson
[{
"name"             : "node\-app",
"cwd"              : "/srv/node\-app/current",
"args"             : ["\-\-toto=heya coco", "\-d", "1"],
"script"           : "bin/app.js",
"node_args"        : ["\-\-harmony", " \-\-max\-stack\-size=102400000"],
"log_date_format"  : "YYYY\-MM\-DD HH:mm Z",
"error_file"       : "/var/log/node\-app/node\-app.stderr.log",
"out_file"         : "log/node\-app.stdout.log",
"pid_file"         : "pids/node\-geo\-api.pid",
"instances"        : 6, //or 0 => 'max'
"min_uptime"       : "200s", // 200 seconds, defaults to 1000
"max_restarts"     : 10, // defaults to 15
"max_memory_restart": "1M", // 1 megabytes, e.g.: "2G", "10M", "100K", 1024... the default unit is byte.
"cron_restart"     : "1 0 * * *",
"watch"            : false,
"ignore_watch"      : ["[\\\\/\\\\\\\\]\\\\./", "node_modules"],
"merge_logs"       : true,
"exec_interpreter" : "node",
"exec_mode"        : "fork",
"env": {
"NODE_ENV": "production",
"AWESOME_SERVICE_API_TOKEN": "xxx"
}
}]
\fR
.RE
.PP
<a name="a988"/>
.SH Schema
.PP
The completely definitions:
.PP
.RS
.nf
{
  "script": {
    "type": "string",
    "require": true
  },
  "args": {
    "type": [
      "array",
      "string"
    ]
  },
  "node_args": {
    "type": [
      "array",
      "string"
    ]
  },
  "name": {
    "type": "string"
  },
  "max_memory_restart": {
    "type": [
      "string",
      "number"
    ],
    "regex": "^\\\\d+(G|M|K)?$",
    "ext_type": "sbyte",
    "desc": "it should be a NUMBER \- byte, \\"[NUMBER]G\\"(Gigabyte), \\"[NUMBER]M\\"(Megabyte) or \\"[NUMBER]K\\"(Kilobyte)"
  },
  "instances": {
    "type": "number",
    "min": 0
  },
  "log_file": {
    "type": [
      "boolean",
      "string"
    ],
    "alias": "log"
  },
  "error_file": {
    "type": "string",
    "alias": "error"
  },
  "out_file": {
    "type": "string",
    "alias": "output"
  },
  "pid_file": {
    "type": "string",
    "alias": "pid"
  },
  "cron_restart": {
    "type": "string",
    "alias": "cron"
  },
  "cwd": {
    "type": "string"
  },
  "merge_logs": {
    "type": "boolean"
  },
  "watch": {
    "type": "boolean"
  },
  "ignore_watch": {
    "type": [
      "array",
      "string"
    ]
  },
  "watch_options": {
    "type": "object"
  },
  "env": {
    "type": ["object", "string"]
  },
  "^env_\\\\S*$": {
    "type": [
      "object",
      "string"
    ]
  },
  "log_date_format": {
    "type": "string"
  },
  "min_uptime": {
    "type": [
      "number",
      "string"
    ],
    "regex": "^\\\\d+(h|m|s)?$",
    "desc": "it should be a NUMBER \- milliseconds, \\"[NUMBER]h\\"(hours), \\"[NUMBER]m\\"(minutes) or \\"[NUMBER]s\\"(seconds)",
    "min": 100,
    "ext_type": "stime"
  },
  "max_restarts": {
    "type": "number",
    "min": 0
  },
  "exec_mode": {
    "type": "string",
    "regex": "^(cluster|fork)(_mode)?$",
    "alias": "executeCommand",
    "desc": "it should be \\"cluster\\"(\\"cluster_mode\\") or \\"fork\\"(\\"fork_mode\\") only"
  },
  "exec_interpreter": {
    "type": "string",
    "alias": "interpreter"
  },
  "write": {
    "type": "boolean"
  },
  "force": {
    "type": "boolean"
  }
}
.fi
.RE
.PP
All the keys can be used in a JSON configured file, and just need to make a small change in CLI, e.g.:
.PP
.RS
.nf
exec_interpreter  \-> \-\-interpreter
exec_mode         \-> \-\-execute_command
max_restarts      \-> \-\-max_restarts
force             \-> \-\-force
.fi
.RE
.PP
Yap, if the \fB\fCalias\fR exists, you can using it as a CLI option, but do not forget to turn the camelCasing to underscore split \- \fB\fCexecuteCommand\fR to \fB\fC\-\-execute_command\fR\&.
.PP
\fBNotes\fP
\- Using quote to make an ESC, e.g.:
.PP
.RS
.nf
  $pm2 start test.js \-\-node\-args "port=3001 sitename='first pm2 app'"
.fi
.RE
.IP
The \fB\fCnodeArgs\fR will be
.PP
.RS
.nf
  [
    "port=3001",
    "sitename=first pm2 app"
  ]
.fi
.RE
.IP
But not
.PP
.RS
.nf
  [
    "port=3001",
    "sitename='first",
    "pm2",
    "app'"
  ]
.fi
.RE
.RS
.IP \(bu 2
RegExp key
.RE
.IP
Matches the keys of configured JSON by RegExp but not a specific String, e.g. \fB\fC^env_\\\\S*$\fR will match all \fB\fCenv\fR keys like \fB\fCenv_production\fR, \fB\fCenv_test\fR, and make sure the values conform to the schemas.
.RS
.IP \(bu 2
Special \fB\fCext_type\fR
.RS
.IP \(bu 2
min_uptime
.RE
.PP
Value of \fB\fCmin_uptime\fR could be:
  \- \fBNumber\fP
    e.g. \fB\fC"min_uptime": 3000\fR means 3000 milliseconds.
  \- \fBString\fP
    In the meantime we are making it briefness and easy configuration: \fB\fCh\fR, \fB\fCm\fR and \fB\fCs\fR, e.g.: \fB\fC"min_uptime": "1h"\fR means one hour, \fB\fC"min_uptime": "5m"\fR means five minutes and \fB\fC"min_uptime": "10s"\fR means ten seconds (At last, it will be transformed into milliseconds).
\- max\fImemory\fPrestart
.PP
Value of \fB\fCmax_memory_restart\fR could be:
  \- \fBNumber\fP
    e.g. \fB\fC"max_memory_restart": 1024\fR means 1024 bytes (\fBNOT BITS\fP).
  \- \fBString\fP

NEWS.md

# News

- 0.7.7
    - Bug fixes
- 0.7.2
    - harmony can be enabled [Enabling harmony](#a66)
    - can pass any options to node via PM2_NODE_OPTIONS, configurable via ~/.pm2/custom_options.sh
    - pid file written in ~/.pm2/pm2.pid
    - startup script support for CentOS
    - --no-daemon option (Alex Kocharin)
    - json file now can be : started/stoped/restarted/deleted
    - coffeescript support for new versions (Hao-kang Den)
    - accept JSON via pipe from standard input (Ville Walveranta)
    - adjusting logical when process got an uncaughtException (Ethanz)
- 0.7.1 integrates hardened reload, graceful reload and strengthened process management

# Updates

## Update from 0.x -> 0.7.2

- CentOS crontab option should not be used anymore and use the new init script with `pm2 startup centos`
- If you use the configuration file or the harmonoy option, you should regenerate the init script


lib/PM2.js
// public api wrapper

var util = require('util')
var EE = require('events').EventEmitter
var CLI = require('./CLI')
var Common = require('./Common')

module.exports = PM2
module.exports.PM2 = PM2
module.exports.PM2App = PM2App

function PM2() {
  if (!(this instanceof PM2)) return new PM2();
  this._connected = false;
  this._connecting = false;
  this._actions = 0;
}

util.inherits(PM2, EE);

PM2.prototype._autoconnect = function(cb) {
  this._actions++;
  if (!this._connected) {
    if (!this._connecting) {
      this._connecting = true;
      CLI.connect(function() {
        this._connecting = false;
        this._connected = true;
        this.emit('connect');
      }.bind(this));
    }
    this.once('connect', cb);
  } else {
    process.nextTick(cb);
  }
  return this;
}

PM2.prototype._autodisconnect = function() {
  this._actions--;
  process.nextTick(function() {
    if (this._actions <= 0 && this._connected && !this._connecting) {
      this._connected = false;
      CLI.disconnect(function() {
        this.emit('disconnect');
      }.bind(this));
    }
  }.bind(this));
  return this;
}

PM2.prototype.app = function(config) {
  var app = PM2App(config);
  var self = this;
  app.start = function(cb) {
    self.start(app._config.script, app._config, function() {
      if (cb) cb.apply(self, arguments)
    })
  }
  return app
}

Object.keys(CLI).forEach(function(name) {
  if (PM2.prototype[name]) return;
  PM2.prototype[name] = function() {
    var args = [], cb;
    for (var i=0; i<arguments.length; i++) {
      args[i] = arguments[i];
    }
    if (typeof(args[i-1]) === 'function') {
      cb = args.pop();
    } else {
      cb = function(){};
    }
    args.push(function() {
      cb.apply(this, arguments);
      this._autodisconnect();
    }.bind(this));

    return this._autoconnect(function() {
      CLI[name].apply(CLI, args);
    });
  }
});

function PM2App(config) {
  if (!(this instanceof PM2App)) return new PM2App(config);
  if (typeof(config) === 'string') config = {
    name: config,
    script: config,
  };
  this._config = config
}

util.inherits(PM2App, EE);

;('node_args name instances error_file out_file pid_file cron_restart' +
' merge_logs watch env run_as_user run_as_group log_date_format' +
' min_uptime max_restarts exec_mode exec_interpreter script').split(' ')
.forEach(function(opt) {
  PM2App.prototype[opt] = function(value) {
    this._config[opt] = value
    return this
  }
});

